### ISSUE TITLE
[Fixed] When performing a search, I receive a 500 internal server error

### ISSUE THREAD

**Issue by valahna** [janeczku/calibre-web#1542]
*Not a bug, a potential solution*

I had this problem for some time, and finally figured out what was going on and why.

**TL;DR** - whenever I searched, I would get a 500. Database disk image is malformed. Rename your metadata.db and have calibre desktop app rebuild the database for you. Happy searching! 👍 

For those who want the entire process/report:

**Describe the bug/problem**
When performing a search, I receive a 500 internal server error.

**To Reproduce**
1. Go to search bar and enter a term to query
2. Click on 'search' or hit enter
3. See error

**returned error**
```
[2020-07-26 00:08:35,596] ERROR {cps:1892} Exception on /search [GET]
Traceback (most recent call last):
File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/result.py", line 1284, in fetchall
l = self.process_rows(self._fetchall_impl())
File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/result.py", line 1230, in _fetchall_impl
return self.cursor.fetchall()
sqlite3.DatabaseError: database disk image is malformed
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
File "/usr/local/lib/python3.6/dist-packages/flask/app.py", line 2447, in wsgi_app
response = self.full_dispatch_request()
File "/usr/local/lib/python3.6/dist-packages/flask/app.py", line 1952, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/usr/local/lib/python3.6/dist-packages/flask/app.py", line 1821, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "/usr/local/lib/python3.6/dist-packages/flask/_compat.py", line 39, in reraise
raise value
File "/usr/local/lib/python3.6/dist-packages/flask/app.py", line 1950, in full_dispatch_request
rv = self.dispatch_request()
File "/usr/local/lib/python3.6/dist-packages/flask/app.py", line 1936, in dispatch_request
return self.view_functions[rule.endpoint](**req.view_args)
File "/app/calibre-web/cps/web.py", line 195, in decorated_view
return func(*args, **kwargs)
File "/app/calibre-web/cps/web.py", line 998, in search
entries = calibre_db.get_search_results(term)
File "/app/calibre-web/cps/db.py", line 581, in get_search_results
)).order_by(Books.sort).all()
File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/query.py", line 3341, in all
return list(self)
File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/loading.py", line 101, in instances
cursor.close()
File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/util/langhelpers.py", line 69, in __exit__
exc_value, with_traceback=exc_tb,
File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/util/compat.py", line 178, in raise_
raise exception
File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/orm/loading.py", line 77, in instances
fetch = cursor.fetchall()
File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/result.py", line 1289, in fetchall
e, None, None, self.cursor, self.context
File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py", line 1512, in _handle_dbapi_exception
sqlalchemy_exception, with_traceback=exc_info[2], from_=e
File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/util/compat.py", line 178, in raise_
raise exception
File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/result.py", line 1284, in fetchall
l = self.process_rows(self._fetchall_impl())
File "/usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/result.py", line 1230, in _fetchall_impl
return self.cursor.fetchall()
sqlalchemy.exc.DatabaseError: (sqlite3.DatabaseError) database disk image is malformed
(Background on this error at: http://sqlalche.me/e/13/4xp6)
```

**Expected behavior**
A list of matching books, authors, and the like returned

**Environment (please complete the following information):**
 - OS: Synology DSM 6.2.3-25426 Update 2
    - Platform | Linux 4.4.59+ #25426 SMP PREEMPT Wed Jul 8 03:21:29 CST 2020 x86_64 x86_64
 - Python version: 3.6.9 (default, Apr 18 2020, 01:56:04) [GCC 8.4.0]
 - Calibre-Web version: 0.6.9 Beta - 1a1d105faee147738071bc44cdb577f72d2e4d6c - 2020-07-23T19:31:48+02:00
 - Docker container: linuxserver/calibre-web
 - Special Hardware: Synology NAS (DS718+)
 - Browser: Any

**Additional context**
This is not a bug or a question, but a proposed solution for others who find themselves facing a similar issue :smile:
1. Shut down calibre-web
2. navigate to your calibre library folder
3. rename metadata.db to something else (I typically use 'metadata.db.bak' or *.old)
4. open calibre (the desktop application/manager).
5. click on library, then 'restore databse'
6. Calibre will restore the entire database. This can take a little while.
7. Once complete, start calibre-web
8. enjoy searching in your library again!

As far as I can tell, my metadata.db file became corrupted at some point while I was adding books and other users were downloading, reading, or some other function was performing. This corrupted the database, and causes the search function (and perhaps some others) to crash. The important thing to look for is in the error message: 'database disk image is malformed'

I am not a DBA by trade, so I was unable to fix this with something like DB browser. I suppose one could, given enough DBA and/or google-fu, but from what I have read, having calibre reconstitute the databse from the metadata it saves throughout your library is potentially faster and easier/more reliable. 
This is the result of taking the risk to add/modify books in calibre whilst calibre-web is running. To avoid this, stopping calibre-web is an easy preventative measure.

**Comment by OzzieIsaacs** [comment]
> I am not a DBA by trade, so I was unable to fix this with something like DB browser.

This is indeed very hard, as the calibre database uses some internal functions you also have to replicate during your repair process. 
The procedure you are describing is very clear and also fully correct from my point of view. The problem could arise if you calibre-web and Calibre access the library at the same time, as you describe it (I can't reproduce it, but also had it once in the last 3 years, but found meanwhile also some bugs). After the database is corrupt all database actions should fail (so viewing all sections where you see book covers, but depending on the problem, it might be possible that only certain pages are not working)
One remark: Calibre is using the metadata.opf files as backup for the database. Calibre-web is currently not using/generating/reading this files, so it's a good idea from time to time running a "metadata backup" in the Calibre desktop program (I don't now the English name, it's in the library maintenance section)

**Comment by OzzieIsaacs** [comment]
Meanwhile the bugs causing this error should be solved
